first-commit-V2 of pe-report src

schmelz21 opened the pull request:
# Re-submitting first-commit and PR #

## 🗣 Original pull request did not align structurally with the python-skeleton repo which in-turn caused issues with the builds/tests. It was easier to create new branch with a new PR. ##

Changes - added 'pe_reports' to 'src' directory and did not modify any other original skeleton code. The intent is to just initialize the repo in-line to dev-standards. Modifications will be made here on out to update specific usage for pe-reports, by modifying the 'examples' code approprietly.

Next Steps - Build out additional code to run report_generation.py which will automatically create 'PE Reports Powerpoint file'  from customer data feeds. Additional actions will include encryption, PDF generation and logic to feed data-frames.

## 🧪 Testing ##

pre-commit test were made locally and no git errors were thrown.
report_generation.py runs successfully - local


## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.

schmelz21 committed:
first-commit-V2 of pe-report src

elr64 commented:
Any idea why check 23/24 is showing as "build Expected — Waiting for status to be reported"

jsf9k commented:
> Any idea why check 23/24 is showing as "build Expected — Waiting for status to be reported"

I think it was an old check that ran at one time but now no longer exists.  I removed it from [the settings page](https://github.com/cisagov/pe-reports/settings/branch_protection_rules/20361733).

dav3r commented on src/pe_reports/data/csv/dhs_ce_df.csv:
```diff
@@ -0,0 +1,2 @@
+﻿count
+11
```
Are all of the CSVs in `src/pe_reports/data/csv/` really needed in source control?  They all appear to be dummy CSVs- there's probably a cleaner solution here.  One possibility is to describe the format of each file in your README.  Another is to put these files in a `samples` or `templates` directory.  Not saying that this is the worst thing in the world, but we generally avoid putting data files (even dummy ones) in source control when we can.

schmelz21 replied:
No these now will be removed.

schmelz21 replied:
Resolved in 7f7de9e

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -0,0 +1,47 @@
+    # Pages.dark_web((prs, data)
+    # Pages.supplimental(prs, data)
+
+    print("\n [Info] File Saved: ../output/Customer_ID_Posture_Exposure.pptx")
```
Is there a good reason to keep this binary output file in source control?

schmelz21 replied:
No, not at this time. It is being removed.

schmelz21 replied:
Resolved in d2d1676

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,149 @@
+"""Need to insert docstring here."""
```
I'd love to see all of these docstrings filled out as part of the initial commit PR.

schmelz21 replied:
Docstrings have been updated.

elr64 replied:
Docstrings resolved in c90e57d

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,149 @@
+from pptx.util import Inches
+from stylesheet import Graph, Paragraph
+
+# Use the folloing fuction to locate "Text Box" element in slide.
```
Typo 😄 
```suggestion
# Use the following function to locate "Text Box" element in slide.
```

dav3r replied:
@schmelz-ctr - When you address/correct something that was commented on in a code review, please reply to the comment with something along the lines of (e.g. for this particular comment):
> Resolved in https://github.com/cisagov/pe-reports/pull/2/commits/c90e57d04d489c77775a0f84c42965ab11184fcd

This helpfully lets the commenter see the exact commit where the relevant change was made and they can then decide if the change is satisfactory or not.  If it is, the person who wrote the original comment will mark the conversation as "Resolved".

Also, if you are unfamiliar with how the "Commit suggestion" feature works in GitHub, I encourage you to [read up on it](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) - it's quite nice.

schmelz21 replied:
Resolved in c90e57d

schmelz21 replied:
Re-Resolved in 7786b61

dav3r reviewed (changes requested):
This is a good start, but generally, the initial-commits PR should include at a minimum the following updates:
* README.md - it doesn't have to be complete, but it should remove all skeleton-specific info and give a basic idea of what the repo is for.
* tests - even if they are simple, they should test your code and not the skeleton code.
* setup.py - customized for your project and all skeleton-specific items removed/updated
* CODEOWNERS - Add an appropriate set of GitHub handles to the [default list of owners](https://github.com/cisagov/skeleton-python-library/blob/develop/.github/CODEOWNERS#L6) (not the `/.github/` owners - that will remain owned by the Fusion Dev team members).

You also should ensure that `src/pe-reports` includes the following important files from the skeleton (customized for your project); you should move those files from the example directory in order to preserve their commit history:
* `__init__.py`
* `__main__.py` 
* `_version.py`

There might be some other items that I missed, but please take care of these first, then re-request reviews and we can go from there.

jsf9k reviewed (changes requested):
I agree with @dav3r's requested changes.  In addition, there is no reason for the `src/example` directory to continue to exist.  You should remove that directory once you have used `git mv` to move the `__init__.py`, `__main__.py`, and `_version.py` files into your `src/pe_reports` directory.

For future reference, it would make more sense to start from the skeleton, do `git mv src/example src/pe_reports`, and then add your source code directly into `src/pe_reports`.

schmelz21 committed:
Updated CODEOWNERS

schmelz21 committed:
Updated docstrings

schmelz21 committed:
Updated python constructs

schmelz21 committed:
Updated setup,py

schmelz21 committed:
Sorted CODEOWNERS list

schmelz21 committed:
Corrected error missing unused __version__

schmelz21 commented:
@dav3r I am getting the expected flake8 error noted in the example __init__.py, which also errors in the checks. Whats the best way to handle this. Am I missing something that was configured in the skeleton or comment it out right now?

pre-commit...
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/pe_reports/__init__.py:6:1: F401 '._version.__version__' imported but unused
src/pe_reports/__init__.py:7:1: F401 '.report_generator.main' imported but unused


i.e -----
We disable a Flake8 check for "Module imported but unused (F401)" here because
although this import is not directly used, it populates the value
package_name.__version__, which is used to get version information about this
Python package.
Third-Party Libraries

schmelz21 committed:
Delete __init__.py

Removed example code.

schmelz21 committed:
Delete __main__.py

Removed example code.

schmelz21 committed:
Delete _version.py

Removed example code.

schmelz21 committed:
Delete example.py

Removed example code.

schmelz21 committed:
Delete secret.txt

Removed example code.

schmelz21 committed:
Update .coveragerc

updated due to test error.

schmelz21 committed:
Delete test_example.py

replacing with test_pe_reports.py

schmelz21 committed:
Add files via upload

Added simple pe_reports test

schmelz21 committed:
Updated siimple test to avoid security issue.

schmelz21 committed:
Updated flake to ignore F401

schmelz21 commented:
Requesting a re-review to align with all minimal requirements for a "first-commit". 
Some quick notes: I updated flake to ignore error F401 per the issue explained and commented on regarding __init__.py. Other issued addressed since last review: Updated CODEOWNERS, setup.py, Docstrings, removal of example code, added in python constructors to src and inserted a simple test and added README.md. Intended next branch and PR will include a fully functional reporting tool for Posture & Exposure.

dav3r commented on setup.py:
```diff
@@ -81,13 +81,23 @@ def get_version(version_file):
     python_requires=">=3.6",
     # What does your project relate to?
-    keywords="skeleton",
+    keywords="Posture and Exposure Reports",
```
This should be a list of lower-case keywords like:
```suggestion
    keywords="posture exposure report",
```

schmelz21 replied:
Resolved in e294e62

dav3r commented on setup.py:
```diff
@@ -42,9 +42,9 @@ def get_version(version_file):
+    name="pe_reports",
     # Versions should comply with PEP440
-    version=get_version("src/example/_version.py"),
+    version=get_version("src/pe_reports/_version.py"),
```
The `description` on the next line should be updated from the skeleton default text.

schmelz21 replied:
Resolved in e294e62

dav3r commented on .flake8:
```diff
@@ -22,4 +22,4 @@ select = C,D,E,F,W,B,B950
 # https://github.com/ambv/black/issues/21. Guido agrees here:
 # https://github.com/python/peps/commit/c59c4376ad233a62ca4b3a6060c81368bd21e85b.
-ignore = E501,W503
+ignore = E501,W503,F401
```
You should remove `F401` from this list- you have globally disabled that Flake8 check for all of your files, which we don't want to do.

Instead, you need to tell Flake8 to ignore that one particular import in `src/pe_reports/__init__.py` [like we do in the skeleton](https://github.com/cisagov/skeleton-python-library/blob/b85e1b85ce7816b8961c0afb2fb9b843ba6e96a5/src/example/__init__.py#L2-L6) by adding `# noqa: F401` to the appropriate line: https://github.com/cisagov/pe-reports/blob/78d906ac241d76e366312eb90548c666321e3b31/src/pe_reports/__init__.py#L6

schmelz21 replied:
Reverted to original state.

schmelz21 replied:
Resolved in bbf7aa7 and in 5c86cbc

dav3r reviewed (changes requested):
You still have some more work to do here - please take a look at my new comments.  

If any previous comments have been corrected, please reply to them and mention the relevant commit(s) so that we can resolve the conversations where appropriate.

dav3r commented:
> @dav3r I am getting the expected flake8 error noted in the example **init**.py, which also errors in the checks. Whats the best way to handle this. Am I missing something that was configured in the skeleton or comment it out right now?

@schmelz-ctr Please see [my reply here](https://github.com/cisagov/pe-reports/pull/2#discussion_r640101466).

dav3r commented:
> I agree with @dav3r's requested changes. In addition, there is no reason for the `src/example` directory to continue to exist. You should remove that directory once you have used `git mv` to move the `__init__.py`, `__main__.py`, and `_version.py` files into your `src/pe_reports` directory.
> 
> For future reference, it would make more sense to start from the skeleton, do `git mv src/example src/pe_reports`, and then add your source code directly into `src/pe_reports`.

@schmelz-ctr - It doesn't look like you did what @jsf9k asked above with regard to using `git mv` for `__init__.py`, `__main__.py`, and `_version.py`.  None of them have their historical commits from the skeleton:
* https://github.com/cisagov/pe-reports/commits/first-commits-V2/src/pe_reports/__init__.py
* https://github.com/cisagov/pe-reports/commits/first-commits-V2/src/pe_reports/__main__.py
* https://github.com/cisagov/pe-reports/commits/first-commits-V2/src/pe_reports/_version.py

(e.g. `__init__.py` should have a commit history that includes these commits from the skeleton: https://github.com/cisagov/skeleton-python-library/commits/develop/src/example/__init__.py)

Please take care of that as well.  If you need additional guidance on how to do this, please reach out to us.

schmelz21 committed:
Revert "Delete _version.py"

This reverts commit 172b730e0ebd02390e31302c33b3482c3d958036.

schmelz21 committed:
Revert "Delete __main__.py"

This reverts commit 517e44900b645f9fb2bf7cad8900719ee0df9d04.

schmelz21 committed:
Revert "Delete __init__.py"

This reverts commit 1d26a0559d4497c72e86ddab8d5d0e08823fe714.

schmelz21 committed:
Moved scripts to example

schmelz21 committed:
Updated constructors wih history

schmelz21 committed:
Deleted bad constructors

schmelz21 committed:
Refresh pe_reports with history

schmelz21 committed:
Added ignore F401 to init

schmelz21 committed:
isort update made

schmelz21 commented:
@dav3r Can we schedule a few minutes on teams to review the "git mv for __init__.py, __main__.py, and _version.py issue". I went through the process this morning, yet somewhere between moving the files from "example" to "pe_reports" and git/github I am loosing the history.
I almost am attempted to re-branch from "develop" and go through the steps requested.

dav3r commented:
> @dav3r Can we schedule a few minutes on teams to review the "git mv for **init**.py, **main**.py, and _version.py issue". I went through the process this morning, yet somewhere between moving the files from "example" to "pe_reports" and git/github I am loosing the history.
> I almost am attempted to re-branch from "develop" and go through the steps requested.

@schmelz-ctr Sure, please ping me on Teams or send me a calendar invite.

schmelz21 committed:
Revert "Added ignore F401 to init"

This reverts commit 7a8f1420945885cb9af209f3e79355615c42d1f9.

schmelz21 committed:
Removed bad pe_report files

schmelz21 committed:
Update constructors with history

schmelz21 committed:
Merge branch 'first-commits-V2' of https://github.com/cisagov/pe-reports into first-commits-V2

schmelz21 commented:
@dav3r as per our conversation last Friday and our attempt to do a `git mv`-[ instructions](https://github.com/cisagov/pe-reports/pull/2#issuecomment-849642568): It appears that even though applying a `git mv` to the src/example files _version.py, __init__.py and __main__.py, tracking history does not display in GitHub. However nativity Git can track history and can be reviewed by using `git log --follow <file>`.
The following 2 links outline this issue: https://stackoverflow.com/questions/2314652/is-it-possible-to-move-rename-files-in-git-and-maintain-their-history and https://git-scm.com/docs/git-log.

Per conversation on next steps, I am cleaning up open items in this PR and will reissue a request for re-review.

schmelz21 committed:
Updated constructors with history

schmelz21 committed:
Revert "Updated constructors wih history"

This reverts commit a4563e9b9e254350caf606ba96f0bc464a553dd0.

schmelz21 committed:
Minor changes addressed

schmelz21 committed:
Delete src/pe_reports/data/csv directory

Removed raw data files

schmelz21 committed:
Delete Customer_ID_Posture_Exposure.pptx

Removed binary, report output

schmelz21 committed:
Linting issue in main

schmelz21 commented:
Requested a re-review with @dav3r.

dav3r commented on setup.py:
```diff
@@ -81,13 +81,23 @@ def get_version(version_file):
+        "docopt",
+        "glob2",
+        "pandas",
+    ],
```
Please sort this list of requirements alphabetically.

schmelz21 replied:
Resolved in e694f6d and e5b92f2

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+from pptx.util import Inches
+from stylesheet import Graph, Paragraph
+
+# Use the following fuction to locate "Text Box" element in slide.
```
You missed a typo here:
```suggestion
# Use the following function to locate "Text Box" element in slide.
```

schmelz21 committed:
Update typo src/pe_reports/pages.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

schmelz21 committed:
sorted setup.py install requirements

schmelz21 committed:
linting issue setup

schmelz21 committed:
Merge branch 'first-commits-V2' of https://github.com/cisagov/pe-reports into first-commits-V2

dav3r commented on README.md:
```diff
@@ -1,25 +1,61 @@
-[pre-commit hooks](https://pre-commit.com) and
-[GitHub Actions](https://github.com/features/actions) configurations
-appropriate for a Python library project.
+This package is used to generate encrypted Posture and Exposure (P&E) pdf
```
We are trying to be better about case-sensitivity for acronyms and such:
```suggestion
This package is used to generate encrypted Posture and Exposure (P&E) PDF
```

schmelz21 replied:
Resolved in 250935b

dav3r commented on README.md:
````diff
@@ -1,25 +1,61 @@
+section of the [CONTRIBUTING](CONTRIBUTING.md) document for information about
+setting up a Python virtual environment.
+
+Required third party libraries: openpyxl, pymongo, pymupdf, pandas, python-pptx
+Install third party libraries using:
+
+```console
+python3 -m pip install LIBRARY
+```
````
Is this section still needed, now that these libraries have been included in setup.py and will get automatically pulled in when this project is installed with pip?
https://github.com/cisagov/pe-reports/blob/22a998a1094aaf507dd1a8c2408930b2d25ae897/setup.py#L90-L99

schmelz21 replied:
Yes, removing this section. At the time it was written these instructions were for the internal PE team prior to any code migration to the skeleton.

schmelz21 replied:
Resolved in 250935b

dav3r commented on setup.py:
```diff
@@ -104,5 +113,5 @@ def get_version(version_file):
         ]
     },
     # Conveniently allows one to run the CLI tool as `example`
```
You missed some skeleton text here:
```suggestion
    # Conveniently allows one to run the CLI tool as `pe-reports`
```

schmelz21 replied:
Resolved in da8dcea

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        slide = prs.slides[1]
+        shape = Paragraph.shapes(prs, slide)
+
+        # Overeview Value - Credentials exposed in recent posts
```
I'm pretty sure this is a typo.  If so, note there are several other occurrences of it below that should be fixed as well.
```suggestion
        # Overview Value - Credentials exposed in recent posts
```

schmelz21 replied:
Ran Code Spell check and updated where appropriate. 

Resolved in 8806233

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -0,0 +1,42 @@
+"""ciagov/pe-reports: A tool for creating Posture & Exposure reports."""
```
Fix a typo by eliminating an unnecessary bit of info:
```suggestion
"""A tool for creating Posture & Exposure reports."""
```

schmelz21 replied:
Resolved in 8806233

dav3r commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+"""Models to manage grahical attributes of presentation layouts, paragraphs, and charts."""
```
Typo:
```suggestion
"""Models to manage graphical attributes of presentation layouts, paragraphs, and charts."""
```

schmelz21 replied:
Resolved in 8806233

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -0,0 +1,33 @@
+        # if both arguments are equal then it's success
+        self.assertEqual("tutorialspoint".upper(), "TUTORIALSPOINT")
+
+    # checking whether a string is upper or not
```
This comment can be removed now that we have docstrings.

schmelz21 replied:
Resolved in 8806233

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class TestingStringMethods(unittest.TestCase):
+    """Simple test for first-commit only."""
```
Since you created dummy tests here, please create an issue in this repo to replace these tests with useful tests.  Then add a `TODO` comment somewhere in this file with a brief note and a link to the issue you created.

schmelz21 replied:
Addressed in 8806233

dav3r replied:
Thank you for creating the issue, though in the future, please fill out our issue template instead of overwriting it with a copy/paste of my comment.

schmelz21 replied:
Got it!

schmelz21 committed:
Update README.md content

schmelz21 committed:
Update typo setup.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

schmelz21 committed:
Updated typos and notation

dav3r reviewed (approved):
@schmelz-ctr - Thanks for sticking with this and working through all of our feedback.  Everything now looks good to me 🎉 , but @jsf9k or @mcdonnnj should also take a look here to ensure that there are no other big things that need to be taken care of at this point.

schmelz21 commented:
> @schmelz-ctr - Thanks for sticking with this and working through all of our feedback. Everything now looks good to me 🎉 , but @jsf9k or @mcdonnnj should also take a look here to ensure that there are no other big things that need to be taken care of at this point.

Dave, thanks. Please reach out if there is anything else you need from me to do.

dav3r commented:
> Dave, thanks. Please reach out if there is anything else you need from me to do.

Once you get your 2nd approval from @jsf9k or @mcdonnnj, you're all clear to merge this PR (you should have the ability to do that, but let me know if you can't).

schmelz21 commented:
> > Dave, thanks. Please reach out if there is anything else you need from me to do.
> 
> Once you get your 2nd approval from @jsf9k or @mcdonnnj, you're all clear to merge this PR (you should have the ability to do that, but let me know if you can't).

Sounds good. Will do.

jsf9k commented on setup.py:
```diff
@@ -42,10 +42,10 @@ def get_version(version_file):
-    version=get_version("src/example/_version.py"),
-    description="Example Python library",
+    version=get_version("src/pe_reports/_version.py"),
+    description="Posture and Exposure Report's library",
```
If the apostrophe is intended to denote plurality instead of ownership then it should be deleted:
```suggestion
    description="Posture and Exposure Reports library",
```

If it is meant to denote ownership then ignore this comment.

schmelz21 replied:
Resolved in 4002c20

jsf9k replied:
Thanks @schmelz-ctr for committing this suggestion.  In the future you might try making the commit message a little more descriptive and specific.  Even the bland and generic "Fix typo" is better than "Update setup.py".

jsf9k commented on README.md:
```diff
@@ -1,25 +1,54 @@
+## Making Changes ##
+
+To change any general report format/standard visuals edit
+/src/data/shell/pe_shell.pptx
+
+To make any style changes, edit /src/pe_reports/stylesheet.py
+
+To change metrics, edit /src/pe_reports/report_metrics.py
+
+To change page setups/graphs, edit /src/pe_reports/pages.py
```
```suggestion
`/src/data/shell/pe_shell.pptx`

To make any style changes, edit `/src/pe_reports/stylesheet.py`

To change metrics, edit `/src/pe_reports/report_metrics.py`

To change page setups/graphs, edit `/src/pe_reports/pages.py`
```

schmelz21 replied:
Resolved in 93db0ef

jsf9k commented on README.md:
```diff
@@ -1,25 +1,54 @@
+
+Required configurations:
+*You must have access to the cyhy database
+Install mongo-db-from-config (<https://github.com/cisagov/mongo-db-from-config>)
```
```suggestion
Install [cisagov/mongo-db-from-config](https://github.com/cisagov/mongo-db-from-config)
```

schmelz21 replied:
Resolved in 93db0ef

jsf9k commented on README.md:
```diff
@@ -1,25 +1,54 @@
-for step-by-step instructions on how to start a new repository from
-a skeleton. This will save you time and effort when configuring a
-new repository!
+Install LibreOffice for powerpoint to pdf conversion
+<https://www.libreoffice.org/get-help/install-howto/macos/>.
```
```suggestion
Install [LibreOffice](https://www.libreoffice.org/get-help/install-howto/macos/) for PowerPoint to PDF conversion.
```

schmelz21 replied:
Resolved in 93db0ef

jsf9k commented on README.md:
```diff
@@ -1,25 +1,54 @@
+
+Install python 3
+
+(Optional) Setting up your Mac:
+<https://github.com/cisagov/development-guide/blob/develop/dev_envs/mac-env-setup.md>
```
```suggestion
(Optional) [Setting up your Mac](https://github.com/cisagov/development-guide/blob/develop/dev_envs/mac-env-setup.md)
```

schmelz21 replied:
Resolved in 93db0ef

jsf9k reviewed (approved):
I had a few minor quibbles that I'd like to see addressed, but nothing that should hold up this PR.  Please take a look at my suggestions.

schmelz21 committed:
Update setup.py

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

schmelz21 committed:
README.md updates

schmelz21 committed:
readme linting

mcdonnnj commented on src/pe_reports/__init__.py:
```diff
@@ -1,9 +1,8 @@
-
-__all__ = ["example_div"]
+# Third-Party Libraries
+from _version import __version__  # noqa: F401
```
This should remain a relative import.

```suggestion
from ._version import __version__ # noqa: F401
```

schmelz21 replied:
Resolved in 44ee65c

mcdonnnj commented on src/pe_reports/__init__.py:
```diff
@@ -1,9 +1,8 @@
-__all__ = ["example_div"]
+# Third-Party Libraries
+from _version import __version__  # noqa: F401
+from report_generator import main  # noqa: F401
```
This needs to be a relative import. Additionally why are you importing this and not adding it to `__all__`? The above case with `__version__` is a special case. This would remove the need for the `noqa` for this import as well.

```suggestion
from .report_generator import main  # noqa: F401
```

schmelz21 replied:
Resolved in 44ee65c

mcdonnnj commented on src/pe_reports/__main__.py:
```diff
@@ -0,0 +1,6 @@
+"""Code to run if this package is used as a Python module."""
+
+# Third-Party Libraries
+from report_generator import main  # noqa: F401
```
This needs to be a relative import. The `noqa` should not be needed because you use this import in this file.

```suggestion
from .report_generator import main
```

schmelz21 replied:
Resolved in 44ee65c

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+class Pages:
+    """Class containing the presentation page information."""
+
+    def cover(prs):
```
This should either have a `@staticmethod` decorator or the first argument should be `self`.

schmelz21 replied:
`@staticmethod` was implemented to do work logically related to the class, but not on objects.
Resolved in 953dc55

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        font = Paragraph.text_style_title(prs, font)
+        return prs
+
+    def overview(prs):
```
This should either have a `@staticmethod` decorator or the first argument should be `self`.

schmelz21 replied:
@staticmethod was implemented to do work logically related to the class, but not on objects.
Resolved in 953dc55

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        text_frame = Paragraph.text_frame(prs, shape)
+        frame = text_frame.paragraphs[0]
+        run = frame.add_run()
+        df_customer = pd.read_csv("src/pe_reports/data/csv/dhs_customer.csv")
```
This file does not exist and this should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        ov_val_01 = Paragraph.text_frame_ov_val(prs, slide, shape, "TextBox 2")
+        frame = ov_val_01.paragraphs[0]
+        run = frame.add_run()
+        df_creds = pd.read_csv("src/pe_reports/data/csv/dhs_creds.csv")
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        ov_val_02 = Paragraph.text_frame_ov_val(prs, slide, shape, "TextBox 82")
+        frame = ov_val_02.paragraphs[0]
+        run = frame.add_run()
+        df_domains = pd.read_csv("src/pe_reports/data/csv/dhs_domains.csv")
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        ov_val_03 = Paragraph.text_frame_ov_val(prs, slide, shape, "TextBox 86")
+        frame = ov_val_03.paragraphs[0]
+        run = frame.add_run()
+        df_malware = pd.read_csv("src/pe_reports/data/csv/dhs_malware.csv")
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        ov_val_04 = Paragraph.text_frame_ov_val(prs, slide, shape, "TextBox 87")
+        frame = ov_val_04.paragraphs[0]
+        run = frame.add_run()
+        df_vulns = pd.read_csv("src/pe_reports/data/csv/dhs_vulns.csv")
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        ov_val_05 = Paragraph.text_frame_ov_val(prs, slide, shape, "TextBox 90")
+        frame = ov_val_05.paragraphs[0]
+        run = frame.add_run()
+        df_web = pd.read_csv("src/pe_reports/data/csv/dhs_web.csv")
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+
+        # Bar Graph - Top Level Domains used for masquerading
+        chart = CategoryChartData()
+        tld_df = pd.read_csv("src/pe_reports/data/csv/dhs_tld_df.csv").fillna(0)
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+
+        # Bar Graph - Sources of credential exposures
+        chart = CategoryChartData()
+        ce_df = pd.read_csv("src/pe_reports/data/csv/dhs_ce_df.csv").fillna(0)
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+
+        # Line Graph - Sources of credential exposures
+        chart = CategoryChartData()
+        web_df = pd.read_csv("src/pe_reports/data/csv/dhs_web_df.csv").fillna(0)
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+        # Bar Graph - Active malware associations
+        chart = CategoryChartData()
+
+        ma_df = pd.read_csv("src/pe_reports/data/csv/dhs_ma_df.csv").fillna(0)
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+
+        # Bar Graph - inferred vulnerabilities found via external observation
+        chart = CategoryChartData()
+        iv_df = pd.read_csv("src/pe_reports/data/csv/dhs_iv_df.csv").fillna(0)
```
This file does not exist and should not be hardcoded like this.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -0,0 +1,42 @@
+import sys
+
+# Third-Party Libraries
+from _version import __version__
```
This needs to be a relative import.

```suggestion
from ._version import __version__
```

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+
+    def shapes(self, slide):
+        """Create a text frame."""
+        for shape in slide.shapes:
+            if not shape.has_text_frame:
+                continue
+        return shape
```
There's no explanation for how reliable just grabbing the first shape that has a text frame is in practice.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+        return shape
+
+    def shapes_find(self, slide):
+        """Find text frames."""
```
There is no reference to text frames in the code for this method.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+    """Simple class to call text frame attributes."""
+
+    def shapes(self, slide):
+        """Create a text frame."""
```
This looks like it's searching for a shape with a text frame.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+        text_frame.clear()
+        return text_frame
+
+    def text_frame_ov_val(self, slide, shape, name):
+        """Text style for overview page values."""
+        for shape in slide.shapes:
```
The `shape` argument is immediately clobbered in this for loop.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+            if shape.name == name:
+                text_frame = shape.text_frame
+                text_frame.clear()
+        return text_frame
```
There is no accounting for a case where there is no `shape` such that `shape.name == name`. This code also implies that the last match for `shape.name == name` has its `.text_frame` element returned and there is no explanation for this behavior.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+                text_frame.clear()
+        return text_frame
+
+    def text_frame_key_metric(self, slide, shape, name):
+        """Text style for metrics."""
+        for shape in slide.shapes:
```
The `shape` argument is immediately clobbered in this for loop.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+            if shape.name == name:
+                text_frame = shape.text_frame
+                text_frame.clear()
+        return text_frame
```
There is no accounting for a case where there is no `shape` such that `shape.name == name`. This code also implies that the last match for `shape.name == name` has its `.text_frame` element returned and there is no explanation for this behavior.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -0,0 +1,100 @@
+
+    def bar(self, slide, chart):
+        """Medium bar chart."""
+        chart.font.size = Pt(10)
+        chart.font.rgb = (20, 200, 100)
```
These values should not be hardcoded like this since they see repeated reuse in this class.

schmelz21 replied:
Resolved in 689417d

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -0,0 +1,42 @@
+
+# Third-Party Libraries
+from _version import __version__
+from pages import Pages
```
This needs to be a relative import.

```suggestion
from .pages import Pages
```

schmelz21 replied:
Resolved in 689417d

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -0,0 +1,150 @@
+from pptx.chart.data import CategoryChartData
+from pptx.enum.chart import XL_CHART_TYPE
+from pptx.util import Inches
+from stylesheet import Graph, Paragraph
```
This needs to be a relative import.

```suggestion
from .stylesheet import Graph, Paragraph
```

schmelz21 replied:
Resolved in 953dc55

mcdonnnj commented on setup.py:
```diff
@@ -81,13 +81,22 @@ def get_version(version_file):
+        "setuptools >= 24.2.0",
+        "pymongo",
+        "pymupdf",
+    ],
```
```suggestion
        "python-pptx",
    ],
```

schmelz21 replied:
Resolved in b25de2a

mcdonnnj commented on setup.py:
```diff
@@ -81,13 +81,22 @@ def get_version(version_file):
     py_modules=[splitext(basename(path))[0] for path in glob("src/*.py")],
     include_package_data=True,
-    install_requires=["docopt", "schema", "setuptools >= 24.2.0"],
+    install_requires=[
```
This entire list should be sorted.

schmelz21 replied:
Resolved in b25de2a

mcdonnnj reviewed (commented):
@schmelz-ctr As this PR was merged this code is unusable. When trying to import:

```console
>>> import pe_reports
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pe-reports/src/pe_reports/__init__.py", line 7, in <module>
    from _version import __version__  # noqa: F401
ModuleNotFoundError: No module named '_version'
```

I have a number of thing that I think should be fixed from this PR, including fixing everything in [this LGTM report](https://lgtm.com/projects/g/cisagov/pe-reports/alerts/?mode=list) I had to fix every relative import and manually install the `python-pptx` package just to successfully perform an `import pe_reports`.

schmelz21 commented:
@mcdonnnj thanks for the feed back. With the PR closed how would you recommend to best update these issues. Should I create a new Pull Request? It does not look like I can amend 'develop'.

dav3r commented:
> @mcdonnnj thanks for the feed back. With the PR closed how would you recommend to best update these issues. Should I create a new Pull Request? It does not look like I can amend 'develop'.

@schmelz-ctr - Yes, please create a new branch and PR to take care of #5 and the other issues that @mcdonnnj mentioned above.  Thanks!

schmelz21 commented:
Sounds good - thanks.
